### PR TITLE
Standardize monster special attack accuracy

### LINKED ISSUE #22161: Taekwondo forces unarmed attacks while using a reach weapon
Game version: 6851

Operating system: Windows 10 64-bit

Tiles or curses: Tiles

Mods active: DDA, Cataclysm++, DinoMod, Deadleaves' Fictional Guns, Medieval Content, Mining Mod, More Survival Tools, No Explosive Zombies, Parks and Rec Buildings, PKs Rebalancing, Packaging, Classic Roguelike Classes, StatsThroughSkills, Tanks and Other Vehicles, ZSFIXED Extra Proffesions, Boats, Crazy Cataclysm, Icecoons Arsenal, Filthy Clothes, Makeshift Items, Classes and Scenerios, Necromancy, Disable NPC Needs, Simple nutrition, Extendfed Realistic Guns

### Expected behavior
While wielding a knife spear I would assume that the reach attack would use the spear, training piercing and melee, While attacking an adjacent enemy would use the forced unarmed attack from Taekwondo, training unarmed and melee.

### Actual behavior
All melee attacks, including reach attacks, are treated as unarmed for damage and skill training while in Taekwondo stance.  Only tested with a knife spear.

### Steps to reproduce the behavior
Equip a reach weapon, only tested knife spear, while using the Taekwondo style. Hit somebody with a reach attack.

### PR DESCRIPTION
**Pull request by Coolthulhu**
Monster special attack accuracy used a weird formula with two separate `rng` calls, making it hard to predict and forcing brute forcing to understand it. When brute forced, it looks rather sigmoidal.
I changed it to use the same formula as melee attack, but with double accuracy. Interestingly, the hit chances are very similar.

Chance to dodge a hulk smash, by player's dodge value (dodge skill + dex/2):
melee-special
Blue is old value, orange is the new one

This makes dodge matter more early on and makes hulks less dangerous to weak characters (good), but prevents very high dodge from providing near-immunity to special attacks (also good). Nowadays medium armors are no longer (as) overpowered, so that extra reward to highly-skilled characters in medium armors shouldn't be a huge deal.

Additionally:
* Instead of having uncanny dodge, dodge check and `on_dodge` as 3 separate lines, they are all merged into `dodge_check`. They (almost) always came together so it's not much of a change
* Dodging monster attacks now costs dodges and is penalized for lack of dodges left in a turn. This is the biggest mechanical change here, hence [CR]
* Fixed and enabled `[melee]` tests

EDIT: Fixes #22161

### CODE REVIEW

File: src/creature.cpp [hunk 1]
```diff
@@ -347,11 +347,16 @@ void Creature::melee_attack(Creature &t, bool allow_special)
     melee_attack( t, allow_special, no_technique_id );
 }
 
+float Creature::hit_roll_at_accuracy( float accuracy )
```
Comment (kevingranade) [inline src/creature.cpp]: Please comment what this means, I suspect it really doesn't mean anything and it needs to be extracted differently.

Comment (Coolthulhu) [reply src/creature.cpp]: It's common math for all melee attacks.
It is necessary to allow different attacks to have different accuracies. `hit_roll` alone is not sufficient, multiplying the result of `hit_roll` by a value would be terrible statistics-wise.

Alternatives are:
* Optional argument for `hit_roll` for different "base accuracy". This one could be beneficial, considering that some day player will need different accuracies for different weapon selections (dual wielding, offhand kicks, headbutt when grappled and unable to use zweihander etc.)
* Inlining the roll directly, replicating the math in the instances of usage of this function
* Decoupling melee math from creature class completely, having the creature only supply accuracy and having the math for attacks be determined by some other class

**Comment by kevingranade** [comment]
Consider extracting the hit_spread calculation into a helper and call it from dodge_check() and melee_actor::call()  instead of injecting the accuracy parameter into creature:deal_melee_attack().
It makes the overal flow a lot simpler.

For hit_roll_at_accuracy(), consider renaming to melee_hit_range(), it at least documents a little bit what it does.

**Comment by i2amroy** [comment]
Overall idea gets a +1 from me, this would definitely be a nice little bit of standardization.

**Comment by Coolthulhu** [comment]
I ran into same issues as here when looking into fixing #22161

> Consider extracting the hit_spread calculation into a helper and call it from dodge_check() and melee_actor::call() instead of injecting the accuracy parameter into creature:deal_melee_attack().
It makes the overal flow a lot simpler.

You mean scrapping the which argument? Accuracy (pre-roll) is needed for `on_dodge` argument, to properly train dodging. `hit_roll` technically isn't 100% necessary, but at the moment it is used to make special attacks train dodge double despite having regular accuracy.

**Comment by Coolthulhu** [comment]
All right, I removed the `hit_spread` argument from the function. Now it only gets the (non-random) accuracy, then rolls that.

Since I needed to do some refactoring anyway, I bundled a fix for #22161 here.
The refactoring involved removing `virtual Creature::hit_roll` and family, then adding it back as non-virtual variants. The one for monsters is identical to old version, the one for characters always needs a weapon supplied. Characters do not have "the" accuracy, only accuracy with a relevant weapon.

**Comment by ZhilkinSerg** [comment]
Marking as Stalled. Feel free to reopen it.

**Pull request closed without merging**